src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -880,8 +880,50 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
                 }
               }
             )
-            else
-              typer1.typed(typer1.applyImplicitArgs(tree), mode, pt)
+            else {
+              if(tree.symbol != null && !tree.symbol.isConstructor) {
+                def stabilizeQual(tree: Tree): Option[(Tree, ValDef)] = tree match {
```

Comment:
Could the transformation be done using this?

```
  case treeInfo.Applied(fun @ Select(qual, name), targs, argss) => ...
```

I also wonder if assigning the types manually to the Select / Apply / TypeApply nodes is necessary. Could we just build the trees and run the typer? Probably the new synthetic value would have to be entered in the typer's scope. Also, the select's symbol might need to be set manually to prevent going into overloading resolution. Similar case for named/default args: https://github.com/scala/scala/blob/v2.13.0-M2/src/compiler/scala/tools/nsc/typechecker/NamesDefaults.scala#L173-L189. See also https://github.com/scala/scala/pull/558.

Reply:
If we have the types available to assign why would we run the typer? Wouldn't that be more expensive?

Reply:
It seems to me that inlining pieces of `typedSelect` and `typedApply` makes the code unnecessarily long / harder to maintain. I don't think we have a performance issue here, we'd just run through a few methods of the typer.